packages/go_router/lib/src/misc/stacked_navigation_shell.dart

```diff
@@ -0,0 +1,221 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:collection/collection.dart';
+import 'package:flutter/widgets.dart';
+
+import '../router.dart';
+import '../state.dart';
+
+/// Transition builder callback used by [StackedNavigationShell].
+///
+/// The builder is expected to return a transition powered by the provided
+/// `animation` and wrapping the provided `child`.
+///
+/// The `animation` provided to the builder always runs forward from 0.0 to 1.0.
+typedef StackedNavigationTransitionBuilder = Widget Function(
+  BuildContext context,
+  Animation<double> animation,
+  Widget child,
+);
+
+/// Builder for the scaffold of a [StackedNavigationShell]
+typedef StackedNavigationScaffoldBuilder = Widget Function(
+  BuildContext context,
+  int currentIndex,
+  List<StackedNavigationItemState> itemsState,
+  Widget scaffoldBody,
+);
+
+/// Representation of a item in the stack of a [StackedNavigationShell]
+class StackedNavigationItem {
+  /// Constructs an [StackedNavigationItem].
+  StackedNavigationItem(
+      {required this.rootRoutePath, required this.navigatorKey});
+
+  /// The location/path of the root route of this navigation tree
+  final String rootRoutePath;
+
+  /// Optional navigatorKey
+  final GlobalKey<NavigatorState> navigatorKey;
+}
+
+/// Represents the current state of a [StackedNavigationItem] in a
+/// [StackedNavigationShell]
+class StackedNavigationItemState {
+  /// Constructs an [StackedNavigationItemState].
+  StackedNavigationItemState(this.item);
+
+  /// The [StackedNavigationItem] this state is representing.
+  final StackedNavigationItem item;
+
+  /// The last location of this item.
+  String? lastLocation;
+
+  /// The [Navigator] for this item.
+  Navigator? navigator;
+
+  /// Gets the current location from the [currentRouterState] or falls back to
+  /// the root route location of the associated [item].
+  String get currentLocation => lastLocation ?? item.rootRoutePath;
+}
+
+/// Widget that maintains a stateful stack of [Navigator]s, using an
+/// [IndexedStack].
+///
+/// Each item in the stack is represented by a [StackedNavigationItem],
+/// specified in the `stackItems` parameter. The stack items will be used to
+/// build the widgets containing the [Navigator] for each index in the stack.
+/// Once a stack item (along with its Navigator) has been initialized, it will
+/// remain in a widget tree, wrapped in an [Offstage] widget.
+///
+/// The stacked navigation shell can be customized by specifying a
+/// `scaffoldBuilder`, to build a widget that wraps the index stack.
+class StackedNavigationShell extends StatefulWidget {
```

Comment:
I will probably keep this private in route.dart. we can expose it later if people find this useful.

Reply:
👍 Done